orangecontrib/bioinformatics/annotation/annotate_samples.py

```diff
@@ -0,0 +1,202 @@
+            Tabular data
+
+        Returns
+        -------
+        :obj:`list`
+            Sets of selected attributes for each cell
+        """
+        if len(data.X) <= 1:
+            return [], []
+        # rank data
+        data_ge_ranked = rankdata(data.X, axis=0)
+
+        # compute U, mu, sigma
+        n = data_ge_ranked.shape[0]
+        n2 = n - 1
+        u = data_ge_ranked - 1
+        mu = n2 / 2
+        sigma = np.zeros(data_ge_ranked.shape[1])
+        for i in range(data_ge_ranked.shape[1]):
+            _, counts = np.unique(data_ge_ranked[:, i], return_counts=True)
+            sigma[i] = np.sqrt(
+                1 * n2 / 12 * ((n + 1) - np.sum((counts ** 3 - counts)) /
+                               (n * (n - 1))))
+
+        # compute z
+        z = (u - mu) / (sigma + 1e-16)
+
+        # gene selection
+        attributes_np = np.array([
+            a.attributes.get("Entrez ID") for a in data.domain.attributes])
+        attributes_sets = [set(attributes_np[row > 1]) - {None} for row in z]
+        return attributes_sets, z
+
+    @staticmethod
+    def _group_marker_attributes(markers):
+        """
+        Function transforms annotations table to dictionary with format
+        {annotation1: [attributes], annotation2: [attributes], ...}
+        """
+        types_dict = defaultdict(set)
+        for m in markers:
+            if m["Entrez ID"].value is not None and \
+                    not m["Entrez ID"].value == "?":
+                types_dict[str(m["Cell Type"])].add(int(m["Entrez ID"].value))
+        return types_dict
+
+    def _assign_annotations(self, items_sets, available_annotations, tax_id):
+        """
+        Function get set of attributes (e.g. genes) that represents each item
+        and attributes for each annotation. It returns the annotations most
+        significant for each cell.
+
+        Parameters
+        ----------
+        items_sets : list of sets
+            Set of most important attributes for each item.
+        available_annotations : Orange.data.Table
+            Available annotations (e.g. cell types)
+
+        Returns
+        -------
+        ndarray
+            Annotation probabilities
+        list
+            Annotation list
+        """
+        # retrieve number of genes for organism
+        N = len(GeneInfo(tax_id))
+
+        grouped_annotations = self._group_marker_attributes(
+            available_annotations)
+        grouped_annotations_items = list(grouped_annotations.items())
+
+        def hg_cell(item_attributes):
+            p_values = []
+            prob = np.zeros(len(grouped_annotations_items))
+            for i, (ct, attributes) in enumerate(grouped_annotations_items):
+                x = len(item_attributes & attributes)
+                k = len(item_attributes)  # drawn balls - expressed for item
+                m = len(attributes)  # marked balls - items for a process
+
+                p_value = self.p_value_fun(x, N, m, k)
```

Comment:
Take a look at the function `score_hypergeometric_test` in `orangecontrib/bioinformatics/utils/statistics.py`. It computes the results using an iterator and might be faster. Can also binomial test be added this way?

Reply:
The function itself cannot be used directly. The map is not faster than for loop using default Python interpreter, maybe it is a bit more readable. I used it anyway, it can be useful in case we would like to use multithreading in the future.